### PR TITLE
Create shared custom view for library items

### PR DESCRIPTION
**Pull request by NotWoods**
Our history and bookmark items are pretty much identical but the code is duplicated. I've rewritten the library item code to based on the bookmarks layout to be shared by all history and bookmark adapters.

The checkmarks have also been changed to use a view switcher, so the background and drawable don't need to be manually updated. This also lets us keep the favicon without reloading when selecting.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

**Comment by colintheshots** [comment]
While I fully support this, can we hold this major change until we branch 1.1?

**Comment by sblatz** [comment]
Bumping @colintheshots on this one :)

File: app/src/main/res/layout/library_site_item.xml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<androidx.constraintlayout.widget.ConstraintLayout
+        xmlns:android="http://schemas.android.com/apk/res/android"
+        xmlns:app="http://schemas.android.com/apk/res-auto"
+        xmlns:tools="http://schemas.android.com/tools"
+        android:layout_width="match_parent"
+        android:layout_height="@dimen/library_item_height"
+        android:background="?android:attr/selectableItemBackground">
+
+    <ImageSwitcher
```
[the remark this answers is not in the document]
Comment (NotWoods) [reply app/src/main/res/layout/library_site_item.xml]: I had plans to animate this in the future, but the image switcher also makes the code much cleaner. Without it we reload the icons on check/uncheck anyways so I think caching the drawable makes more sense.

File: app/src/main/java/org/mozilla/fenix/library/LibrarySiteItemView.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    fun changeSelected(isSelected: Boolean) {
+        icon.displayedChild = if (isSelected) 1 else 0
+    }
+
+    suspend fun loadFavicon(url: String) = withContext(IO) {
+        context.components.core.icons.loadIcon(IconRequest(url)).await().bitmap
+    }
+
+    enum class ItemType {
+        SITE, FOLDER, SEPARATOR;
```
Comment (boek) [inline app/src/main/java/org/mozilla/fenix/library/LibrarySiteItemView.kt]: is SEPARATOR used anywhere?

File: app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkAdapter.kt [hunk 3]
```diff
@@ -226,82 +192,49 @@ class BookmarkAdapter(val emptyView: View, val actionEmitter: Observer<BookmarkA
...
+                actionEmitter.onNext(action)
             }
 
-            bookmark_layout.setOnLongClickListener {
+            view.setOnLongClickListener {
                 if (mode == BookmarkState.Mode.Normal) {
-                    if (selected) actionEmitter.onNext(BookmarkAction.Deselect(item)) else actionEmitter.onNext(
-                        BookmarkAction.Select(item)
-                    )
+                    actionEmitter.onNext(BookmarkAction.Select(item))
```
[the remark this answers is not in the document]
Comment (NotWoods) [reply app/src/main/java/org/mozilla/fenix/library/bookmarks/BookmarkAdapter.kt]: There is actually no state where the bookmark is in normal mode and selected, so I'm just removing dead code.

File: app/src/main/java/org/mozilla/fenix/library/LibrarySiteItemView.kt (hunk 2 above)
Comment (NotWoods) [reply app/src/main/java/org/mozilla/fenix/library/LibrarySiteItemView.kt]: It's used by separators in the bookmarks fragment.

**Pull request merged**